### PR TITLE
Unified Syscalls

### PR DESCRIPTION
**Pull request by rennergade**
This is the ultimate merge of the past several months work of getting rid of the alternate "lind_syscalls" pathways in libc and nacl.

Let's all give it a final review before we merge it in.

### CODE REVIEW

File: src/trusted/service_runtime/nacl_syscall_common.c [hunk 1]
```diff
@@ -4550,3 +4542,968 @@ int32_t NaClSysSigProcMask(struct NaClAppThread *natp, int how, const void *set,
...
+    goto cleanup;
+  }
+
+  NaClFastMutexLock(&nap->desc_mu);
+  for(unsigned int i = 0; i < nfds; ++i) {
+      
+    if((lind_fds[i].fd = descnum2Lindfd(nap, fds[i].fd)) < 0) {
+      NaClLog(2, "NaClSysPoll was passed an unrecognized file descriptor, returning %d\n", fds[i].fd);
+      return fds[i].fd;
+    }
```
Comment (jesings) [inline src/trusted/service_runtime/nacl_syscall_common.c]: Memory leak, this return doesn't go to cleanup

File: src/trusted/service_runtime/nacl_syscall_common.c [hunk 2]
```diff
@@ -4550,3 +4542,968 @@ int32_t NaClSysSigProcMask(struct NaClAppThread *natp, int how, const void *set,
...
+    
+  hd->d = ret; //old NaClHostDescCtor 
+  hd->flags = NACL_ABI_O_RDWR; //old NaClHostDescCtor 
+  
+  hd->cageid = nap->cage_id;
+  code = NaClSetAvail(nap, ((struct NaClDesc *) NaClDescIoDescMake(hd)));
+  userfd = NextFd(nap->cage_id);
+  fd_cage_table[nap->cage_id][userfd] = code;
+  code = userfd;
+  //Postprocessing end
```
Comment (jesings) [inline src/trusted/service_runtime/nacl_syscall_common.c]: remove these *processing comments, they're just clutter at this point

File: src/trusted/service_runtime/nacl_syscall_common.c [hunk 3]
```diff
@@ -4550,3 +4542,968 @@ int32_t NaClSysSigProcMask(struct NaClAppThread *natp, int how, const void *set,
...
+      return (long) safeposixreadfds;
+    }
+  }
+  
+  if(writefds) {
+    naclwritefds = (fd_set*) NaClUserToSysAddrRange(nap, (uintptr_t) writefds, sizeof(fd_set));
+
+    if ((void*) kNaClBadAddress == naclwritefds) {
+      NaClLog(2, "NaClSysBind could not translate write fds address, returning %d\n", -NACL_ABI_EFAULT);
+      return -NACL_ABI_EFAULT;
```
Comment (jesings) [inline src/trusted/service_runtime/nacl_syscall_common.c]: memory leak here for safeposixreadfds

File: src/trusted/service_runtime/nacl_syscall_common.c [hunk 4]
```diff
@@ -4550,3 +4542,968 @@ int32_t NaClSysSigProcMask(struct NaClAppThread *natp, int how, const void *set,
...
+      goto cleanup;
+    }
+  }
+
+  if(exceptfds) {
+    naclexceptfds = (fd_set*) NaClUserToSysAddrRange(nap, (uintptr_t) exceptfds, sizeof(fd_set));
+
+    if ((void*) kNaClBadAddress == naclexceptfds) {
+      NaClLog(2, "NaClSysBind could not translate except fds address, returning %d\n", -NACL_ABI_EFAULT);
+      return -NACL_ABI_EFAULT;
```
Comment (jesings) [inline src/trusted/service_runtime/nacl_syscall_common.c]: memory leak here for safeposix read/write fds

File: src/trusted/service_runtime/nacl_syscall_common.c [hunk 5]
```diff
@@ -4550,3 +4542,968 @@ int32_t NaClSysSigProcMask(struct NaClAppThread *natp, int how, const void *set,
...
+      goto cleanup;
+    }
+  }
+
+  if(timeout) {
+    nacltimeout = (struct timeval*) NaClUserToSysAddrRange(nap, (uintptr_t) timeout, sizeof(struct timeval));
+
+    if ((void*) kNaClBadAddress == nacltimeout) {
+      NaClLog(2, "NaClSysBind could not translate timeout address, returning %d\n", -NACL_ABI_EFAULT);
+      return -NACL_ABI_EFAULT;
```
Comment (jesings) [inline src/trusted/service_runtime/nacl_syscall_common.c]: memory leak...

File: src/trusted/service_runtime/nacl_syscall_common.c [hunk 6]
```diff
@@ -4550,3 +4542,968 @@ int32_t NaClSysSigProcMask(struct NaClAppThread *natp, int how, const void *set,
...
+
+void fd_set_fd_translator_fromlind(struct NaClApp* nap, fd_set *fdset, char* otherbitfield, int maxfd) {
+  int fds[FD_SETSIZE];
+  int fdsindex = 0;
+
+  for(int i = 0; i < maxfd; i++) {
+    if(FD_ISSET(i, fdset)) {
+      int translated_fd = fds[fdsindex++] = descnum2Lindfd(nap, i);
+      NaClLog(LOG_FATAL, "User fd %d which was valid on entering select call invalid upon exit\n", i);
+      if(!(otherbitfield[translated_fd / 8] & 1 << (translated_fd & 7))) {
```
Comment (jesings) [inline src/trusted/service_runtime/nacl_syscall_common.c:5387]: This bit shift should be parenthesized

File: src/trusted/service_runtime/nacl_syscall_common.c [hunk 7]
```diff
@@ -4550,3 +4542,968 @@ int32_t NaClSysSigProcMask(struct NaClAppThread *natp, int how, const void *set,
...
+}
+
+void fd_set_fd_translator_fromlind(struct NaClApp* nap, fd_set *fdset, char* otherbitfield, int maxfd) {
+  int fds[FD_SETSIZE];
+  int fdsindex = 0;
+
+  for(int i = 0; i < maxfd; i++) {
+    if(FD_ISSET(i, fdset)) {
+      int translated_fd = fds[fdsindex++] = descnum2Lindfd(nap, i);
+      NaClLog(LOG_FATAL, "User fd %d which was valid on entering select call invalid upon exit\n", i);
```
Comment (jesings) [inline src/trusted/service_runtime/nacl_syscall_common.c:5386]: This should be surrounded by an `if(translated_fd < 0)`

**Review by jesings — COMMENTED** [review]
Mostly looks good, I think we all got a little sloppy by the end in our rush to get this out, myself included, so some changes still need to be made

**Review by kapkic — APPROVED** [review]
.

**Review by youngfalcon — COMMENTED** [review]
Looks great in general. The code is more readable than ever. However, the memory leaks comments pointed out in other reviews seem problematic and require attention.

**Comment by rennergade** [comment]
I agree with the above comments. Let's each fix the sections we contributed in the first place.

**Pull request merged**